README.md
# Eclipse Keyple website  

Repository to the jeanpierrefortune.github.io/keyple-website website of the 'Eclipse Keyple' project.

The Eclipse Keyple website is generated with Hugo.

## Getting started

### Local installation
Install [hugo extended](https://gohugo.io/getting-started/installing/) on your machine.
Install dependencies, build assets and start a webserver:
```bash
hugo server
```

Checkout http://localhost:1313/

### Using docker 

You can use the image "ext-alpine" from the project : https://github.com/klakegg/docker-hugo. An "ext" is needed image because it includes golang.


Launch a shell into hugo container from the root directory of this project
```bash
docker run --rm -it \
  -v $(pwd):/src \
    -p 1313:1313 \
  klakegg/hugo:ext-alpine \
  shell
```

Then run
```bash
hugo server --environment development
```

If you want to avoid downloading modules at each run, install them locally by running
```bash
hugo mod vendor
```

## Related projects

### [Wowchemy](https://github.com/wowchemy/wowchemy-hugo-modules)

[Wowchemy theme](https://wowchemy.com/) for [Hugo](https://gohugo.io/).

releases/releases.md

---

{{% release-row "2023-05-10" "Keyple Card Calypso Java Lib" "2.3.5" %}} 
### Fixed
- An issue with "Select File" and "Read Record" commands when the card is a legacy ASK Tango card having the following startup information value:
  - `03 08 03 04 00 02 00`
### Upgraded
- "Keyple Service Resource Library" to version `2.1.1`
{{% /release-row %}}

---

{{% release-row "2023-05-05" "Keyple Distributed Remote Java Lib" "2.2.1" %}} 
### Fixed
- Fixes the communication issue between client and server components when using the "Reader Client Side" usage mode 
  that appeared with version `2.2.0`.
{{% /release-row %}}

---

{{% release-row "2023-05-05" "Keyple Service Java Lib" "2.2.1" %}} 
### Fixed
- Fixes the communication issue between client and server components of the "Distributed" solution that appeared with version `2.1.4`.
{{% /release-row %}}

---

{{% release-row "2023-05-05" "Keyple Util Cpp Lib" "2.3.0.4" %}} 

{{% /release-row %}}

---

{{% release-row "2023-05-05" "Keyple Card Calypso Cpp Lib" "2.2.5.1" %}} 

{{% /release-row %}}

---

{{% release-row "2023-05-05" "Keyple Plugin Pcsc Cpp Lib" "2.1.0.1" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-27" "Keyple Plugin Cardresource Java Lib" "1.0.1" %}} 
### Fixed
- Exception handling associated with the reader allocation process.
### Upgraded
- "Keyple Service Resource Library" to version `2.1.1`
{{% /release-row %}}

---

{{% release-row "2023-04-27" "Keyple Service Resource Java Lib" "2.1.1" %}} 
### Fixed
- The issue of resource allocation associated with regular (non-pool) plugins, happening when the resource usage timeout was not explicitly configured.
{{% /release-row %}}

---

{{% release-row "2023-04-25" "Keyple Plugin Cardresource Java Lib" "1.0.0" %}} 
This is the initial release.
{{% /release-row %}}

---

{{% release-row "2023-04-25" "Keyple Service Resource Java Lib" "2.1.0" %}} 
### Added
- The method `CardResource.getReaderExtension` to access the reader extension.
### Changed
- `CardResource` is now an interface.
### Upgraded
- "Calypsonet Terminal Reader API" to version `1.2.0`
- "Keyple Service Library" to version `2.2.0`
- "Keyple Util Library" to version `2.3.0`
- "Google Gson Library" (com.google.code.gson) to version `2.10.1`
{{% /release-row %}}

---

{{% release-row "2023-04-25" "Keyple Service Java Lib" "2.2.0" %}} 
:warning: **CAUTION**: In the case of a distributed mode, it is required to use at least this version on the client and server side.
### Added
- The `PoolPlugin.getSelectedSmartCard` method to retrieve the smart card if it has been automatically selected by the reader allocation process.
### Upgraded
- "Keyple Plugin API" to version `2.1.0`
{{% /release-row %}}

---

{{% release-row "2023-04-25" "Keyple Plugin Java Api" "2.1.0" %}} 
### Added
- The method `PoolPluginSpi.allocateReader` now returns the new interface `PoolReaderSpi`, which extends `ReaderSpi`, in
  order to retrieve the smart card that may be automatically selected.
- `CHANGELOG.md` file (issue [eclipse/keyple#6]).
- CI: Forbid the publication of a version already released (issue [#7]).
{{% /release-row %}}

---

{{% release-row "2023-04-24" "Keyple Plugin Pcsc Java Lib" "2.1.2" %}} 
### Fixed
- Unnecessary logging in Java 16+ during reflexive calls used to overcome the deficiencies of `smartcard.io` in Windows 
environment.
{{% /release-row %}}

---

{{% release-row "2023-04-17" "Keyple Util Cpp Lib" "2.3.0.3" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-12" "Keyple Service Cpp Lib" "2.1.1" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-12" "Keyple Card Calypso Cpp Lib" "2.2.5" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-12" "Keyple Cpp Example" "2022-11-17" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-12" "Keyple Util Cpp Lib" "2.3.0.2" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-12" "Keyple Card Calypso Cpp Lib" "2.2.4" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-12" "Keyple Card Calypso Cpp Lib" "2.2.3" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-09" "Keyple Cpp Example" "2022-10-27" %}} 

{{% /release-row %}}

---

{{% release-row "2023-04-05" "Keyple Plugin Pcsc Java Lib" "2.1.1" %}} 
### Changed
- `PcscPluginFactoryBuilder.useContactReaderIdentificationFilter` method marked as deprecated.
### Fixed
- Use default reader type identification filters.
- Logging of the available readers names in the `AbstractPcscPluginAdapter` class.
{{% /release-row %}}

---

{{% release-row "2023-04-04" "Keyple Distributed Local Java Lib" "2.2.0" %}} 
:warning: **CAUTION**: this version requires to use at least version `2.2.0` of the
[Keyple Distributed Remote Library](https://keyple.org/components-java/distributed/keyple-distributed-remote-java-lib/)!
### Added
- `CHANGELOG.md` file (issue [eclipse/keyple#6]).
- CI: Forbid the publication of a version already released (issue [#5]).
### Changed
- Initial card content and user input/output data used for "ReaderClientSide" mode are now serialized/de-serialized 
  as JSON objects, and no more as strings containing JSON objects.
- All JSON property names are now "lowerCamelCase" formatted.
### Upgraded
- "Keyple Distributed Network Library" to version `2.2.0`.
- "Keyple Util Library" to version `2.3.0`.
- "Google Gson Library" (com.google.code.gson) to version `2.10.1`.
{{% /release-row %}}

---

{{% release-row "2023-04-04" "Keyple Distributed Remote Java Lib" "2.2.0" %}} 
:warning: **CAUTION**: this version requires to use at least version `2.2.0` of the
[Keyple Distributed Local Library](https://keyple.org/components-java/distributed/keyple-distributed-local-java-lib/)!
### Changed
- Initial card content and user input/output data used for "ReaderClientSide" mode are now serialized/de-serialized
  as JSON objects, and no more as strings containing JSON objects.
### Upgraded
- "Keyple Distributed Network Library" to version `2.2.0`.

.github/scripts/generate_all.sh
#!/bin/bash

# Check if the repos.list file exists
if [ ! -f ./.github/scripts/repos.list ]; then
  echo "The repos.list file cannot be found."
  exit 1
fi

# Remove all .md files in the current directory
echo "Removing existing .md files..."
rm -f ./*.md

# Loop through each line in the repos.list file
while read -r repo; do
  if [ ! -z "$repo" ]; then
    echo "Processing repository: $repo"
    ./get_releases.sh "$repo"
  fi
done < repos.list

# Concatenate all .md files into a new releases.md file, sorted in reverse alphabetical order by their namess
echo "Concatenating .md files into releases.md..."
touch releases.md
for md_file in $(ls ./*.md | sort -r); do
  cat "$md_file" >> releases.md
  echo -e "\n---\n" >> releases.md
done

echo "All .md files have been concatenated into releases.md."


.github/scripts/get_releases.sh
#!/bin/bash

# Module name
# Replace '-' with ' '
temp="${1//-/ }"

# Capitalize the first letter of each word
module_name=""

for word in $temp; do
    capitalized_word="$(tr '[:lower:]' '[:upper:]' <<< "${word:0:1}")${word:1}"
    module_name+="$capitalized_word "
done

# Download and process JSON data
curl -s -S -H "Accept: application/vnd.github.v3+json" "https://api.github.com/repos/eclipse/$1/releases?per_page=1000&page=1" | \
jq -c '.[] | tojson' | while read -r i; do
  i=$(echo "$i" | jq -r '.')
  tag_name=$(echo "$i" | jq -r '.tag_name')
  published_at=$(echo "$i" | jq -r '.published_at')
  body=$(echo "$i" | jq -r '.body' | sed 's/\\r\\n/\n/g')
  filename="$published_at.md"
  # Write the Markdown file contents
  echo "{{% release-row \"$(echo "$published_at" | cut -c 1-10)\" \"${module_name% }\" \"$tag_name\" %}} " > "$filename"
  echo -e "$body" >> "$filename"
  echo "{{% /release-row %}}" >> "$filename"
done

.github/scripts/generate_all_releases.sh
#!/bin/bash

mkdir releases/
cd releases || exit

# Check if the repos.list file exists
if [ ! -f ../../.github/scripts/repos.list ]; then
  echo "The repos.list file cannot be found."
  exit 1
fi

# Remove all .md files in the current directory
echo "Removing existing .md files..."
rm -f ./*.md

# Loop through each line in the repos.list file
while read -r repo; do
  if [ ! -z "$repo" ]; then
    echo "Processing repository: $repo"
    ../../.github/scripts/get_releases.sh "$repo"
  fi
done < ../../.github/scripts/repos.list

# Concatenate all .md files into a new releases.md file, sorted in reverse alphabetical order by their namess
echo "Concatenating .md files into releases.md..."
touch releases.md
for md_file in $(ls ./*.md | sort -r); do
  cat "$md_file" >> releases.md
  echo -e "\n---\n" >> releases.md
done

echo "All .md files have been concatenated into releases.md."


.github/scripts/generate_changelog_page.sh
#!/bin/bash

# Check if the repos.list file exists
if [ ! -f ./.github/scripts/repos.list ]; then
  echo "The repos.list file cannot be found."
  exit 1
fi

# Remove all .md files in the current directory
echo "Removing existing .md files..."
rm -f ./*.md

# Loop through each line in the repos.list file
while read -r repo; do
  if [ ! -z "$repo" ]; then
    echo "Processing repository: $repo"
    ./.github/scripts/get_releases.sh "$repo"
  fi
done < ./.github/scripts/repos.list

# Concatenate all .md files into a new changelog file, sorted in reverse alphabetical order by their names
echo "Concatenating .md files into changelog..."
touch changelog
for md_file in $(ls ./*.md | sort -r); do
  cat "$md_file" >> changelog
done

awk -v replacement="$(cat ./changelog)" '{gsub("TO_BE_REPLACED", replacement)}1' content/community/changelog/_index.md > temp.md && mv temp.md content/community/changelog/_index.md


.github/scripts/dashboard_check_repos_status_curl.sh
#!/bin/sh

# This script checks if the target repository has been updated since the last check.
# returns 0 if true, 1 if false.

token=$1
repository_name=$2

filter_website_repository=""

if [ "$repository_name" = "keyple-website" ]; then
  filter_website_repository=" -e \"pushed_at\""
fi

echo "Checking repository "$repository_name"..."

# Compare hash value of the API result by excluding these fields:
# - "size": Unknown reason of size changing.
# - "pushed_at": Cyclic commit when gh-pages branch is update (keyple-website).

github_json=`curl --request GET \
          --url https://api.github.com/repos/eclipse/$repository_name \
          --header "authorization: Bearer $token" \
          --header "content-type: application/json" | grep -v -e "size"$filter_website_repository`
github_hash=`echo $github_json | md5sum | cut -d ' ' -f 1`

dashboard_json=`curl --request GET \
          --url https://jeanpierrefortune.github.io/keyple-website/dashboard/$repository_name"_.json" \
          --header "content-type: application/json" | grep -v -e "size"$filter_website_repository`
dashboard_hash=`echo $dashboard_json | md5sum | cut -d ' ' -f 1`

if [ "$github_hash" = "$dashboard_hash" ]; then
  echo "No update required"
  exit 1
else
  echo "Update required"
  echo "################################################################################"
  echo "     Github JSON hash = "$github_hash
  echo "  Dashboard JSON hash = "$dashboard_hash
  echo "###################"
  echo "##  Github JSON  ##"
  echo "###################"
  echo $github_json
  echo "######################"
  echo "##  Dashboard JSON  ##"
  echo "######################"
  echo $dashboard_json
  echo "################################################################################"
  exit 0
fi

assets/js/custom.js
// make the footer full width
let pageFooterContainer = $(".page-footer>.container");
pageFooterContainer.addClass("container-fluid");
pageFooterContainer.removeClass("container");
pageFooterContainer.css("padding-left", "0");
pageFooterContainer.css("padding-right", "0");

// Copy to clipboard
updateClipboard = function(newClip, button) {
    navigator.permissions.query({name: "clipboard-write"}).then(result => {
        if (result.state == "granted" || result.state == "prompt") {
            /* write to the clipboard now */
            navigator.clipboard.writeText(newClip).then(function() {
                /* clipboard successfully set */
                button.innerHTML = "Copied";
                setTimeout(function() {
                    button.innerHTML = "Copy";
                }, 2000);
            }, function() {
                /* clipboard write failed */
                alert("Error during the copy to clipboard process!");
            });
        } else {
            alert("Unable to copy to clipboard because your browser does not have the permissions to use the \"clipboard-write\" API!");
        }
    });
}
copyTabContentToClipboard = function(tabpaneId, button) {
    let tabId = $("#"+tabpaneId+" a.active")[0].getAttribute("aria-controls");
    let content = $("#"+tabId+" code")[0].innerText;
    updateClipboard(content, button);
}
copyCodeContentToClipboard = function(codeId, button) {
    let content = $("#code-"+codeId)[0].innerText;
    updateClipboard(content, button);
}

// Dynamic dependencies for "client application" profile of "configuration wizard" java page
const appDependencies = new Set();
computeAppContent = function(language) {
    let tagPrefix = 'code#all-'+language+'-dependencies span#';
    let contentHtml = $(tagPrefix+'calypsonet-terminal-reader-java-api').html();
    if (appDependencies.has("cardCalypso")) {
        contentHtml += "\n" + $(tagPrefix+'calypsonet-terminal-calypso-java-api').html();
    }
    if (appDependencies.has("cardCalypsoLegacySam")) {
        contentHtml += "\n" + $(tagPrefix+'calypsonet-terminal-calypso-crypto-legacysam-java-api').html();
    }
    contentHtml += "\n" + $(tagPrefix+'keyple-common-java-api').html()
        + "\n" + $(tagPrefix+'keyple-util-java-lib').html()
        + "\n" + $(tagPrefix+'keyple-service-java-lib').html();
    if (appDependencies.has("serviceResource") || appDependencies.has("pluginCardResource")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-service-resource-java-lib').html();
    }
    if (appDependencies.has("distributedLocal") || appDependencies.has("distributedRemote")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-distributed-network-java-lib').html();
    }
    if (appDependencies.has("distributedLocal")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-distributed-local-java-lib').html();
    }
    if (appDependencies.has("distributedRemote")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-distributed-remote-java-lib').html();
    }
    if (appDependencies.has("cardGeneric")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-card-generic-java-lib').html();
    }
    if (appDependencies.has("cardCalypso")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-card-calypso-java-lib').html();
    }
    if (appDependencies.has("cardCalypsoLegacySam")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-card-calypso-crypto-legacysam-java-lib').html();
    }
    if (appDependencies.has("pluginAndroidNfc")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-plugin-android-nfc-java-lib').html();
    }
    if (appDependencies.has("pluginAndroidOmapi")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-plugin-android-omapi-java-lib').html();
    }
    if (appDependencies.has("pluginCardResource")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-plugin-cardresource-java-lib').html();
    }
    if (appDependencies.has("pluginPcsc")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-plugin-pcsc-java-lib').html();
    }
    if (appDependencies.has("pluginStub")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-plugin-stub-java-lib').html();
    }
    return contentHtml;
}
updateAppDependencies = function(tabRef, checkbox) {
    if (checkbox != null) {
        if (checkbox.checked == true) {
            appDependencies.add(checkbox.id);
        } else {
            appDependencies.delete(checkbox.id);
        }
    }
    let tabContentId = "#tabs-"+tabRef+"-content";
    $(tabContentId+" .language-gradle")[0].innerHTML = computeAppContent("groovy");
    $(tabContentId+" .language-kotlin")[0].innerHTML = computeAppContent("kotlin");
    $(tabContentId+" .language-xml")[0].innerHTML = computeAppContent("maven");
}

// Dynamic dependencies for "custom Keyple card extension" profile of "configuration wizard" java page
const cardDependencies = new Set();
computeCardContent = function(language) {
    let tagPrefix = 'code#all-'+language+'-dependencies span#';
    let contentHtml = $(tagPrefix+'calypsonet-terminal-reader-java-api').html()
        + "\n" + $(tagPrefix+'calypsonet-terminal-card-java-api').html()
        + "\n" + $(tagPrefix+'keyple-common-java-api').html()
        + "\n" + $(tagPrefix+'keyple-util-java-lib').html();
    if (cardDependencies.has("cardServiceResource")) {
        contentHtml += "\n" + $(tagPrefix+'keyple-service-resource-java-lib').html();
    }
    return contentHtml;
}
updateCardDependencies = function(tabRef, checkbox) {
    if (checkbox != null) {
        if (checkbox.checked == true) {
            cardDependencies.add(checkbox.id);
        } else {
            cardDependencies.delete(checkbox.id);
        }
    }
    let tabContentId = "#tabs-"+tabRef+"-content";
    $(tabContentId+" .language-gradle")[0].innerHTML = computeCardContent("groovy");
    $(tabContentId+" .language-kotlin")[0].innerHTML = computeCardContent("kotlin");
    $(tabContentId+" .language-xml")[0].innerHTML = computeCardContent("maven");
}

// Data table of dependency check tool
initDatatableDependencyCheck = function() {
    $('#datatable-dependency-check').DataTable( {
        ordering: false,
        paging: false,
        initComplete: function () {
            // Remove the search input text & the footer information showing the number of results
            $("#datatable-dependency-check_wrapper div.row:nth-child(3)").remove();
            $("#datatable-dependency-check_wrapper div.row:nth-child(1)").remove();
            // Add the select button to the header of each column
            this.api().columns().every( function () {
                var column = this;
                var select = $('<select><option value=""></option></select>')
                    .appendTo( $(column.header()) )
                    .on( 'change', function () {
                        var val = $.fn.dataTable.util.escapeRegex(
                            $(this).val()
                        );
                        column
                            .search( val ? '^'+val+'$' : '', true, false )
                            .draw();
                    } );
                column.data().unique().sort().reverse().each( function ( d, j ) {
                    select.append( '<option value="'+d+'">'+d+'</option>' )
                } );
            } );
        }
    } );
}

// Load the project dashboard table content
loadProjectDashboard = async function() {

    let rootUrl = window.location.href + "../";

    if (rootUrl.includes("localhost")) {
        rootUrl = "https://jeanpierrefortune.github.io/keyple-website/";
    }

    async function getJson(fileName) {
        let response = await fetch(rootUrl + 'dashboard/' + fileName + '.json?date='+Date.now());
        return await response.json();
    }

    async function getJsonRepositoryData(repository_name, resource) {
        let response = await fetch(rootUrl + 'dashboard/' + repository_name + resource + '.json?date='+Date.now());
        return await response.json();
    }

    async function getReposData(rowIndex, owner, project) {

        // create row
        let newRow = body.insertRow();
        newRow.id = rowIndex;

        const json = await getJsonRepositoryData(project[0], "_");

        // new row
        const row = document.getElementById(rowIndex);

        // column repos name
        let cell = row.insertCell(-1);
        let a = document.createElement('a');
        let linkText = document.createTextNode(json.name);
        a.appendChild(linkText);
        a.title = json.name;
        a.href = json.html_url;
        a.target = "_blank";
        cell.appendChild(a);

        // column doc
        cell = row.insertCell(-1);
        if (project[1] === true) {
            cell.setAttribute("class", "text-center");
            let space = document.createTextNode("  ");
            cell.appendChild(space);
            a = document.createElement('a');
            cell.appendChild(space);
            a.innerHTML = "<i class=\"fas fa-book\"></i>";
            a.title = "API documentation for " + json.name;
            a.href = "https://eclipse.github.io/" + json.name;
            a.target = "_blank";
            cell.appendChild(a);
        }

        // column repos status
        cell = row.insertCell(-1);
        cell.setAttribute("id", "repos-status-" + rowIndex);
        cell.setAttribute("class", "text-center");

        // column latest release
        cell = row.insertCell(-1);
        cell.setAttribute("id", "latest-release-" + rowIndex);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(""));

        // column release date
        cell = row.insertCell(-1);
        cell.setAttribute("id", "release-date-" + rowIndex);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(""));

        // column latest tag
        cell = row.insertCell(-1);
        cell.setAttribute("id", "latest-tag-" + rowIndex);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(""));

        // column issues
        cell = row.insertCell(-1);
        cell.setAttribute("id", "issue-" + rowIndex);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(json.open_issues_count.toString()));
        if (json.open_issues_count > 0) {
            cell.style.backgroundColor = "orange";
        }

        // column branches
        cell = row.insertCell(-1);
        cell.setAttribute("id", "branch-" + rowIndex);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(""));

        // column pull requests
        cell = row.insertCell(-1);
        cell.setAttribute("id", "pull-" + rowIndex);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(""));

        // column forks
        cell = row.insertCell(-1);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(json.forks.toString()));

        // column stars
        cell = row.insertCell(-1);
        cell.setAttribute("class", "text-center");
        cell.appendChild(document.createTextNode(json.stargazers_count.toString()));

        // column updated
        cell = row.insertCell(-1);
        cell.setAttribute("class", "text-center");
        cell.innerHTML = formatDate(json.pushed_at);

        // column created
        cell = row.insertCell(-1);
        cell.setAttribute("class", "text-center");
        cell.innerHTML = formatDate(json.created_at);

        // get complementary data
        await getBranches(rowIndex, owner, project[0]);
        await getPullData(rowIndex, owner, project[0]);
        await getLatestRelease(rowIndex, owner, project[0]);
        await getReleaseDate(rowIndex, owner, project[0]);
        await getLatestTag(rowIndex, owner, project[0]);
        if (project[2] === true) {
            await getStatus(rowIndex, owner, project[0]);
        }

        // debug
        console.log(json.data);
    }

    async function getBranches(rowIndex, owner, repos) {
        const json = await getJsonRepositoryData(repos, "_branches");
        let cell = document.getElementById("branch-" + rowIndex);
        cell.innerHTML = '' + json.length;
        if (json.length > 2) {
            cell.style.backgroundColor = "#FFE0F0";
        }
    }

    async function getPullData(rowIndex, owner, repos) {
        const json = await getJsonRepositoryData(repos, "_pulls");
        let cell = document.getElementById("pull-" + rowIndex);
        if (json.length > 0) {
            cell.style.backgroundColor = "orange";
        }
        cell.innerHTML = json.length;
    }

    async function getLatestRelease(rowIndex, owner, repos) {
        let cell = document.getElementById("latest-release-" + rowIndex);
        try {
            const json = await getJsonRepositoryData(repos, "_releases_latest");
            if (json.tag_name !== undefined) {
                cell.innerHTML = json.tag_name;
            }
        } catch (err) {
        }
    }

    async function getReleaseDate(rowIndex, owner, repos) {
        let cell = document.getElementById("release-date-" + rowIndex);
        try {
            const json = await getJsonRepositoryData(repos, "_releases_latest");
            if (json.tag_name !== undefined) {
                cell.innerHTML = formatDate(json.published_at);
            }
        } catch (err) {
        }
    }

    async function getLatestTag(rowIndex, owner, repos) {
        let cell = document.getElementById("latest-tag-" + rowIndex);
        try {
            const json = await getJsonRepositoryData(repos, "_tags");
            cell.innerHTML = json[0].name;
        } catch (err) {
        }
    }

    async function getStatus(rowIndex, owner, repos) {

        let cell = document.getElementById("repos-status-" + rowIndex);
        let json;
        let branch;

        try {
            json = await getJsonRepositoryData(repos, "_commits_status");
            branch = "main";
        } catch (err) {
        }

        const a = document.createElement('a');
        const linkText = document.createTextNode("\u2b24");
        a.appendChild(linkText);
        a.title = "CI status page";
        a.href = "https://ci.eclipse.org/keyple/job/Keyple/job/" + repos + "/job/" + branch + "/";
        a.target = "_blank";

        switch (json.state) {
            case "error":
            case "failure":
                a.style.color = "red";
                a.title += ": failure";
                break;
            case "pending":
                a.style.color = "orange";
                a.title += ": pending";
                break;
            case "success":
                a.style.color = "green";
                a.title += ": success";
                break;
        }
        cell.appendChild(a);
    }

    function formatDate(dateString) {
        let d = new Date(dateString),
            month = '' + (d.getMonth() + 1),
            day = '' + d.getDate(),
            year = d.getFullYear(),
            hour = '' + d.getHours(),
            min = '' + d.getMinutes(),
            sec = '' + d.getSeconds();

        if (month.length < 2)
            month = '0' + month;
        if (day.length < 2)
            day = '0' + day;
        const date = [year, month, day].join('-');

        if (hour.length < 2)
            hour = '0' + hour;
        if (min.length < 2)
            min = '0' + min;
        if (sec.length < 2)
            sec = '0' + sec;

        const time = [hour, min, sec].join(':');

        return '<span data-toggle="tooltip" title="' + date + ' ' + time + '">' + date + '</span>';
    }

    let owner = "eclipse";

    const lastUpdate = await getJson('datetime');
    const date = new Date(lastUpdate.datetime);

    const offset = date.getTimezoneOffset();
    const isoLocalDate = new Date(date.getTime() - (offset*60*1000));

    const dateOptions = {hour: '2-digit', minute:'2-digit', hour12: false, timeZoneName: 'short'};
    $("#project-dashboard-datetime")[0].innerHTML = isoLocalDate.toISOString().split('T')[0] + ", " + date.toLocaleTimeString('en-EN', dateOptions);

    let projects = await getJson('repository_list');

    // create promises
    const body = document.getElementById("project-dashboard-content");
    let promises = [];
    for (let i = 0; i < projects.length; i++) {
        let promise = getReposData((i + 1).toString(), owner, projects[i])
        promises.push(promise);
    }

    await (async () => {
        await Promise.all(promises)
            .finally(function () {
                $('#project-dashboard-table').DataTable({
                    "lengthMenu": [25, 50, 75, 100],
                    "pageLength": 50,
                    "order": [[11, 'desc']],
                    "oLanguage": {"sSearch": "Filter:"}
                });
                $('.dataTables_length').addClass('bs-select');
                // update the container's width with the real table size
                $('.universal-wrapper').width($('#project-dashboard-table')[0].scrollWidth);
            });
    })();
}
